### PR TITLE
Add a request to list all the tests within a workspace or document

### LINKED ISSUE #611: Add custom request to return a list of all the tests in a project with locations
This is something that would be used by the Visual Studio Code extension to provide location information for tests in the Test Explorer. 

Currently the extension uses the `swift test --list-tests` command to get a list of tests, but this doesn't return locations of the tests. I did add a request https://github.com/apple/swift-package-manager/issues/5601 to add this information to the `--list-tests` output. But it seems it has been decided it would be easier to do this using SourceKit-LSP.

### PR DESCRIPTION
**Pull request by ahoppen**
Requires https://github.com/apple/indexstore-db/pull/175

---

Fixes #611
rdar://98710526

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
A couple things I didn’t consider in the initial draft:
- I think I didn’t include transitive subclasses of `XCTest`
- I also included test methods that had parameters
- Apparently, things work completely differently on Linux and Windows and subclasses of `XCTestCase` aren’t actually modeled as such in the index. Still need to investigate this one.

**Comment by grynspan** [comment]
Does this work with swift-testing too?

File: Sources/SourceKitLSP/SourceKitServer.swift [hunk 1]
```diff
@@ -1490,33 +1494,30 @@ extension SourceKitServer {
...
+    let symbols = findWorkspaceSymbols(matching: req.query).map(WorkspaceSymbolItem.init)
     return symbols
   }
 
+  func workspaceTests(_ req: WorkspaceTestsRequest) async throws -> [WorkspaceSymbolItem]? {
+    let testSymbols = workspaces.flatMap { (workspace) -> [SymbolOccurrence] in
+      guard let index = workspace.index else {
+        return []
+      }
+      let xctestSubclasses = index.occurrences(ofUSR: "c:objc(cs)XCTest", roles: [.baseOf])
```
Comment (stmontgomery) [inline Sources/SourceKitLSP/SourceKitServer.swift:1506]: The superclass to look for is named "XCTest**Case**". The class named plain "XCTest" is its superclass, but it's abstract and we only search for subclasses of the former.

**Comment by ahoppen** [comment]
https://github.com/apple/indexstore-db/pull/175

@swift-ci Please test

**Comment by ahoppen** [comment]
https://github.com/apple/indexstore-db/pull/175

@swift-ci Please test

**Comment by ahoppen** [comment]
https://github.com/apple/indexstore-db/pull/175

@swift-ci Please test Windows

**Comment by ahoppen** [comment]
https://github.com/apple/indexstore-db/pull/175

@swift-ci Please test Windows

**Comment by ahoppen** [comment]
https://github.com/apple/indexstore-db/pull/175

@swift-ci Please test

**Comment by ahoppen** [comment]
https://github.com/apple/indexstore-db/pull/175

@swift-ci Please test

**Comment by ahoppen** [comment]
https://github.com/apple/indexstore-db/pull/175

@swift-ci Please test

**Comment by ahoppen** [comment]
https://github.com/apple/indexstore-db/pull/175

@swift-ci Please test

**Comment by ahoppen** [comment]
https://github.com/apple/indexstore-db/pull/175

@swift-ci Please test

**Comment by ahoppen** [comment]
https://github.com/apple/indexstore-db/pull/175

@swift-ci Please test Windows

**Comment by ahoppen** [comment]
https://github.com/apple/indexstore-db/pull/175

@swift-ci Please test Windows

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Pull request merged**